ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift

```diff
@@ -294,18 +294,65 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
       self.swallowBackspaceTextChange = false
       return
     }
-    
+
     let contextBeforeInput = textDocumentProxy.documentContextBeforeInput ?? ""
     let selection = textDocumentProxy.selectedText ?? ""
     let contextAfterInput = textDocumentProxy.documentContextAfterInput ?? ""
     let context = "\(contextBeforeInput)\(selection)\(contextAfterInput)"
-    let bLength = contextBeforeInput.utf16.count
-    let sLength = selection.utf16.count
+    let bLength = contextBeforeInput.count
+    let sLength = selection.count
     setContextState(text: context, range: NSMakeRange(bLength, sLength))
     // Within the app, this is triggered after every keyboard input.
     // We should NOT call .resetContext() here for this reason.
   }
-  
+
+  // Pre-condition:  no text is selected.  As this is currently only called by `insertText`
+  // below, this condition is met.
+  func sendContextUpdate() {
+    let preCaretContext = textDocumentProxy.documentContextBeforeInput ?? ""
+    let postCaretContext = textDocumentProxy.documentContextAfterInput ?? ""
+
+    let updater = { (_ before: String, _ after: String) -> Void in
+      let contextWindowText = "\(before)\(after)"
+
+      // On the KMW app/webview side, `location` is SMP-aware.
+      // .utf16:  treats SMPs as two char.  (Tested with U+1D5BA)
+      // Without .utf16 - treats SMPs as one char.
+      let range = NSRange(location: before.count, length: 0)
+
+      os_log("setContextState(text: `%@`, range initializer location: %d, range.location: %d", log: KeymanEngineLogger.engine, type: .debug, contextWindowText, before.count, range.location)
+      self.setContextState(text: contextWindowText, range: range)
+    }
+
+    updater(preCaretContext, postCaretContext)
+
+    if preCaretContext == "" {
+      /* The `textDocumentProxy` abstraction is documented (in passing) as involving
+       * inter-process communication.  It is thus asynchronous.  Despite all attempts to prod
```

Comment:
Source: https://developer.apple.com/documentation/uikit/keyboards_and_input/creating_a_custom_keyboard#3570264

> Custom keyboards run in an isolated process that doesn’t have direct access to the text input view. [...]

Also https://developer.apple.com/documentation/uikit/keyboards_and_input/creating_a_custom_keyboard/handling_text_interactions_in_custom_keyboards#overview.

Note:  neither is the actual documentation for `TextDocumentProxy`.